Add in blog post about sbt BSP (#2193)
scalameta/metals: website/blog/2020-11-06-sbt-BSP-support.md

---
author: Chris Kipp
title: sbt BSP support
authorURL: https://twitter.com/ckipp01

> authorImageURL: https://avatars1.githubusercontent.com/u/13974112?s=400&u=7b6a2ddab8eec6f99e4e40ae9b81f71cb5ba92e5&v=4

Author: Huh, funny enough `https://github.com/ckipp01.png` just redirects to `https://avatars2.githubusercontent.com/u/13974112?v=4`. So I can at least shorten it to that.

---

If you've been following the [sbt
releases](https://github.com/sbt/sbt/releases), 1.4.x introduced some great new
features, one of those being BSP support. This effort was initiated by community
members and proposed to the [Scala Center Advisory

> Board.](https://github.com/scalacenter/advisoryboard/blob/master/proposals/023-bsp.md).

Author: 👍 ha yes, trying to do my duty as a good board member and bring attention to it.

Then thanks to the work of [Adrien Piquerez](https://twitter.com/adrienpi2) and
[Eugene Yokota ](https://twitter.com/eed3si9n) BSP support became a reality in
1.4.0. You can read a more detailed account of the reason behind the work and

> some of the details of the implementation in this blog post, [BSP Support in sbt

Author: 👍 I'll take the `the`, but replace the `,` with a `:` which is more fitting I think.

1.4](https://www.scala-lang.org/blog/2020/10/27/bsp-in-sbt.html).

> The Metals team is also happy to announce that as of 0.9.5, the process to use

Author: Ah, I was sort of thinking they would be so close together it didn't matter, but no worries. I changed it to `upcoming 0.9.5 release`

> The Metals team is also happy to announce that as of 0.9.5, the process to use

Author: Actually thinking about this a bit more, these instructions mention the new `generate-bsp-config` which users won't have until the new release of the VS Code extension, so maybe we shouldn't advertise this too much before the actual release to avoid confusion? wdyt?

> The Metals team is also happy to announce that as of 0.9.5, the process to use

Reviewer: The release notes will come up only after the VS Code extension is also released, so it should be fine.

> The Metals team is also happy to announce that as of 0.9.5, the process to use

Reviewer: We can also try to maybe publish it at the same time then? No need for the upcoming in that case.

> The Metals team is also happy to announce that as of 0.9.5, the process to use

Author: Sounds good 👍

other BSP servers like sbt is now much smoother. In some ways, BSP is an
implementation detail abstracted away from a user, and you may not directly
interact at all with your build server, even though it's pivotal for the Metals
experience. Therefore, we'd like to answer some common questions about what this
means to you as a user, explain some of the default choices Metals makes for
you, and show you how to explore the sbt BSP server.

## What is BSP?

First off, you may be wondering what BSP is. The [Build Server
Protocol](https://build-server-protocol.github.io/) (BSP) is a _Protocol for IDEs and
build tools to communicate about compilation, running, testing, debugging, and
much more._ If you're familiar with the [Language Server
Protocol](https://microsoft.github.io/language-server-protocol/) (LSP), BSP is
complementary to LSP and inspired by it. Where LSP allows your editor to
abstract over various programming languages by a shared way to communicate to a
language server, BSP allows IDE's to abstract over various build servers. A
typical example of this can be illustrated like so:

> ![LSP BSP example](https://i.imgur.com/0RRUDlU.png)

Reviewer: Could you maybe add titles to each icon?

Author: Yea sure, I added them in before each image.

You have your editor (Emacs) communicating with Metals via LSP, and then Metals
communicating with a BSP server (Bloop) via BSP. This communication over BSP can
be about compiling your code, running your code, defining sources in your
workspace, etc. You can read all about the various communication types [here in
the
protocol](https://build-server-protocol.github.io/docs/specification#actual-protocol).

[Bloop](https://scalacenter.github.io/bloop/) was the first server to implement
BSP, and it's the default build server for all build tools that Metals supports

> at the moment. When you open a fresh project, you're prompted to import you

Reviewer 2: ```suggestion
at the moment. When you open a fresh project, you're prompted to import your
```

build, and this import process is running a form of "Bloop Install" to write
your build definition to `.bloop/` for Bloop to read and use. Then as you

> continue to edit diagnostics and other information are flowing back and forth

Reviewer 2: ```suggestion
continue to edit, diagnostics and other information are flowing back and forth
```

from Metals to Bloop. Hopefully this gives you a brief picture of what BSP is,
and how it's used in Metals.

## What does sbt BSP support mean for Metals?

> Another question you may have is "what does sbt BSP support" mean for Metals?

Reviewer 2: The question is quoted, so I would add the "mean for Metals" part to it.

```suggestion
Another question you may have is "what does sbt BSP support mean for Metals"?
```

> This means a couple different things. Up until this point if you wanted to use

Reviewer 2: ```suggestion
This means a couple different things. Up until this point, if you wanted to use
```

sbt BSP, you needed to clear your `.metals/` and `.bloop` directories, and then
start sbt with a specific flag before connecting to it. None of this necessary
anymore. In order for [BSP
discovery](https://build-server-protocol.github.io/docs/server-discovery.html)
to happen, you need a `.bsp/*json` file with instructions on how to
start/connect to the build server. A new command has been added to Metals (which

> we'll go over down below) that can generate this for you if it doesn't exist,

Reviewer: ```suggestion
we'll go over down below) that can generate this file for you if it doesn't exist,
```

> and then you will be automatically connected to sbt. If this already exist, then

Reviewer: ```suggestion
and then you will be automatically connected to sbt. If this file already exist, then
```

> we provide a simple way for you to "switch" build servers and for you choice to

Reviewer: ```suggestion
we provide a simple way for you to "switch" build servers and for your choice to
```

be remembered. We also automatically include an sbt plugin to ensure the correct
semantic information is produced by the compiler. So to summarize, Metals now
has the ability to generate the necessary `.bsp/sbt.json` file if it doesn't
exist, and to also switch back and forth from using Bloop or sbt as your build
server.

## What's the difference between Bloop and sbt BSP?

> At this point you may be asking, "whats the difference?". For an average user,

Reviewer 2: ```suggestion
At this point you may be asking, "what's the difference?". For an average user,
```

there may not be a ton of difference, however I'd like to outline a couple
things that may be relevant to you.

- Bloop supports the [Debug Adapter
    Protocol](https://microsoft.github.io/debug-adapter-protocol/) (DAP) and sbt
    doesn't. So you'll notice when using sbt as a build server, you won't have
    the `run` or `debug` code lenses directly in your editor like you do with
    Bloop.
- Bloop writes all the necessary information about your build it needs to disk
    whereas sbt loads it into memory. If you're a user that works for long
    periods of time on a single project, then this may not matter to you at all.
    However, if you jump in and out of projects, without having sbt shell
    running, then you will pay the cost of loading up your build every time.

> - Bloop offers sbt file support for completions and hover, whereas this is not

Reviewer: ```suggestion
- Bloop offers sbt file support that enables completions and hover, whereas this is not
```

    yet available in the sbt BSP implementation.
- sbt gives you access to the full task graph, so for example if you're using
    `BuildInfo`, and you compile with Bloop, your `BuildInfo` won't get
    generated. However, sources will get generated when using sbt server. This
    can help avoid any potential inconsistencies. 

While there are some more differences, these are probably the main ones you'll
notice as a user. These are both great tools, and you'll have to explore more to
see what is the best fit for you. For now, Metals defaults to using Bloop mainly
because we believe it provides a richer feature set and because pretty much all
of our testing includes Bloop as a build server.

## How do I try it out?

Now to the good stuff. There are two main ways that you can start using sbt
server in Metals.

### No `.bsp/sbt.json` exists

> If no `.bsp/sbt.json` exists yet, you can generate it the new

Reviewer: ```suggestion
If no `.bsp/sbt.json` exists yet, you can generate it with the new
```

then auto-connect to the sbt build server. For example in VS Code this looks
like this:

![generate-bsp-config](https://i.imgur.com/kBNbtzI.gif)

### `.bsp/sbt.json` already exists

If you've ran sbt on your project already, your `.bsp/sbt.json` file will
already exist. In this case, you can simply use the `metals.bsp-switch` command
and choose sbt. For example, using `coc-metals`, it looks like this:

![bsp-switch](https://i.imgur.com/6tY2ofL.gif)

### Switching back to Bloop

After you do a `metals.bsp-switch` or `metals-generate-bsp-config`, your choice
will be remembered. So the next time that you open your workspace, you will
automatically connect to the last build server you had chosen. If you'd like to
go back to your previous build server, you can simply use the
`metals.bsp-switch` command again to choose your previous server. For editors
that have an html doctor, you can also reset your build server choice in the
Doctor.

![Doctor](https://i.imgur.com/YEGfEGB.png)

## Conclusion

We hope you enjoy this easier way to use sbt server with Metals. If you're
curious about the implementation details, much of the work was done in [this

> pr](https://github.com/scalameta/metals/pull/2154).  As always please don't

Reviewer: ```suggestion
pr](https://github.com/scalameta/metals/pull/2154). As always please don't
```

> hesitate to ask questions on our various channels, submit issues, or create new

Reviewer: ```suggestion
hesitate to ask questions on our various channels, submit issues or create new
```

I did comment on this in a couple of places, but maybe it's my bad knowledge of the English language. Do we need a coma before `and` or `or` ?

Author: Welcome to the oxford comma debate. https://www.grammarly.com/blog/oxford-comma-debate/

In short, both are correct 😆

Reviewer: Och no... :(
